### DISCUSSION TITLE
map/parJoin

### DISCUSSION THREAD

**Discussion by Fristi** [typelevel/fs2#2225] [Q&A] [answered]
I've got this stream where we unzip files from a incoming Stream:

This works

```scala
archive
  //unzip
  .through(Archive.unzip(blocker))
  .parEvalMapUnordered(5) { entry =>
    entry.body.through(store.put(Path(s"import/$id/${entry.name}"))).compile.drain.as(entry.name)
  }
  .compile
  .toList
```

While this doesn't

```scala
archive
  //unzip
  .through(Archive.unzip(blocker))
  .map(entry => entry.body.through(store.put(Path(s"import/$id/${entry.name}"))).as(entry.name))
  .parJoin(5)
  .compile
  .toList
```

With the latter, it uploads the file to the blob store, but it doesn't yield the `entry.name`

**Comment by mpilquist** [comment, marked as answer]
Check the implementation of `store.put` -- I'm guessing it doesn't emit anything. E.g., `def put(...): Stream[F, Unit] = Stream.eval_(takeAction(...))` emits no output.

**Comment by Fristi** [comment]
Ah that's correct :) My guess that `  .map(entry => entry.body.through(store.put(Path(s"import/$id/${entry.name}"))) >> Stream.emit(entry.name))` would work, but it yields the same result

**Reply by SystemFw** [reply]
you need to use `++` , not `>>`

**Reply by Fristi** [reply]
That was my guess aswell! Was missing the `.drain` on the non-emitting stream.

```scala
archive
  //unzip
  .through(Archive.unzip(blocker))
  .map(entry => entry.body.through(store.put(Path(s"import/$id/${entry.name}"))).drain ++ Stream.emit(entry.name))
  .parJoin(5)
  .compile
  .toList
```

That works 🎉
